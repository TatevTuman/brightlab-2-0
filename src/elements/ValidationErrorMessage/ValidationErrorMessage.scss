.validation-error {
  /* You can describe here your own error message type like in Button.tsx*/
  color: var(--danger);
}
