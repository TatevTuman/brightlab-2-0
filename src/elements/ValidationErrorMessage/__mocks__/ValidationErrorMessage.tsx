import { ValidationErrorMessageProps } from '@elements'

export default (props: ValidationErrorMessageProps) => null
