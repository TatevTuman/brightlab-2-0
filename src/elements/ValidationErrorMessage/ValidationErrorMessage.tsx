import React, { memo } from 'react'
import { useFormContext } from 'react-hook-form'
import { ErrorMessage } from '@hookform/error-message'
import { Children } from '@types'
import './ValidationErrorMessage.scss'

export type ValidationErrorMessageType = { message: string }
export interface ValidationErrorMessageProps {
  name: string
  render?(error: ValidationErrorMessageType): Children
}

/*
  Validation error component for react-hook-form.
  Remember that you can use this component to render anything you want when field is not valid.
  You can render Modal if validation won't complete. If you want to use this component in a different way -
  create another component with different name. For example ValidationConfirmPassword component opens a modal window
  to ask an admin password to do a high risk operation in the app.
*/

/*
  Multiple Error Messages Examaple

  <ValidationErrorMessage
    errors={errors}
    name="multipleErrorInput"
    render={({ messages }) =>
      messages &&
      Object.entries(messages).map(([type, message]) => (
        <p key={type}>{message}</p>
      ))
    }
  />
*/

const ValidationErrorMessage: React.FC<ValidationErrorMessageProps> = props => {
  const { name, render } = props
  const formContext = useFormContext()

  if (!formContext) return null

  /* Gets error from errors object by name and passes it in render */
  const renderErrorMessage = (error: ValidationErrorMessageType) => {
    if (render) return render(error)
    else {
      return (
        <div className={'validation-error'} role="validation-error" data-testid="validation-error">
          {error.message}
        </div>
      )
    }
  }

  return <ErrorMessage errors={formContext.errors} name={name} render={renderErrorMessage} />
}

export default memo(ValidationErrorMessage)
