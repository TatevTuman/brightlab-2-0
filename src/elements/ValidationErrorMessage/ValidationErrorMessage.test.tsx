import React from 'react'
import { cleanup, render, waitFor } from '@testing-library/react'
import ValidationErrorMessage, { ValidationErrorMessageType } from './ValidationErrorMessage'

jest.useFakeTimers()

// beforeAll(() => {})
// afterAll(() => {})
// beforeEach(() => {})
afterEach(() => {
  cleanup()
  jest.clearAllMocks()
})

describe('ValidationErrorMessage', () => {
  const props = {
    errors: {
      test: {
        type: 'required',
        message: 'Test is required'
      }
    },
    name: 'test'
  }

  it('renders correctly', async () => {
    const { container, getByText } = render(<ValidationErrorMessage {...props} />)
    await waitFor(() => container)

    expect(container).toMatchSnapshot()
    expect(getByText(props.errors.test.message)).toBeInTheDocument()
  })

  it('uses render prop', async () => {
    const renderProp = (error: ValidationErrorMessageType) => (
      <div role={'validation-error'} data-testid={'validation-error'}>
        {error.message}
      </div>
    )
    const { getByTestId } = render(<ValidationErrorMessage {...props} render={renderProp} />)

    expect(getByTestId('validation-error')).toBeInTheDocument()
  })
})
