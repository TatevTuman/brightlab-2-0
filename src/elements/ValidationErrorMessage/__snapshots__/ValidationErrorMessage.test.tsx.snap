// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ValidationErrorMessage renders correctly 1`] = `<div />`;
